mypy/checkexpr.py

```diff
@@ -1840,89 +1828,234 @@ def check_op_local(self, method: str, base_type: Type, arg: Expression,
+            # In all other cases, we do the usual thing and call __add__ first and
+            # __radd__ second when doing "A() + B()".
+
+            variants_raw.append((left_op, left_type, right_expr))
+            variants_raw.append((right_op, right_type, left_expr))
+
+        # STEP 2b:
+        # When running Python 2, we might also try calling the __cmp__ method.
+
+        is_python_2 = self.chk.options.python_version[0] == 2
+        if is_python_2 and op_name in nodes.ops_falling_back_to_cmp:
+            cmp_method = nodes.comparison_fallback_method
+            left_cmp_op = lookup_operator(cmp_method, left_type)
+            right_cmp_op = lookup_operator(cmp_method, right_type)
+
+            if bias_right:
+                variants_raw.append((right_cmp_op, right_type, left_expr))
+                variants_raw.append((left_cmp_op, left_type, right_expr))
+            else:
+                variants_raw.append((left_cmp_op, left_type, right_expr))
+                variants_raw.append((right_cmp_op, right_type, left_expr))
+
+        # STEP 3:
+        # We now filter out all non-existant operators. The 'variants' list contains
+        # all operator methods that are actually present, in the order that Python
+        # attempts to invoke them.
+
+        variants = [(op, obj, arg) for (op, obj, arg) in variants_raw if op is not None]
+
+        # STEP 4:
+        # We now try invoking each one. If an operation succeeds, end early and return
+        # the corresponding result. Otherwise, return the result and errors associated
+        # with the first entry.
+
+        errors = []
+        results = []
+        for method, obj, arg in variants:
+            local_errors = make_local_errors()
+
+            callable_name = None  # type: Optional[str]
+            if isinstance(obj, Instance):
+                # TODO: Find out in which class the method was defined originally?
+                # TODO: Support non-Instance types.
+                callable_name = '{}.{}'.format(obj.type.fullname(), op_name)
+
+            result = self.check_call(method, [arg], [nodes.ARG_POS],
+                                   context, arg_messages=local_errors,
+                                   callable_name=callable_name, object_type=obj)
+            if local_errors.is_errors():
+                errors.append(local_errors)
+                results.append(result)
+            else:
+                return result
+
+        # STEP 4b:
+        # Sometimes, the variants list is empty. In that case, we fall-back to attempting to
+        # call the __op__ method (even though it's missing).
+
+        if len(errors) == 0:
+            local_errors = make_local_errors()
+            result = self.check_op_local(op_name, left_type, right_expr, context, local_errors)
+
+            if local_errors.is_errors():
+                errors.append(local_errors)
+                results.append(result)
+            else:
+                return result
```

Comment:
I am not sure I understand how can this ever succeed, could you please add a comment with an example?

Reply:
I originally added that out just in case, but it seems we do fall into that case at least once -- for example, adding an assert there makes [this test case fail](https://gist.github.com/Michael0x2a/0822390ad828329d80400eb959b5d7d6).

I added a TODO.